libsh/docs/libError.man
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "LIBERROR.SH 1"
.TH LIBERROR.SH 1 "2017-02-11" "perl v5.18.2" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "Name"
.IX Header "Name"
libError
.SH "Description"
.IX Header "Description"
Error handling and debug library. To use it just source it on the begining of your script.
.SH "Dependencies"
.IX Header "Dependencies"
no dependencies.
.SH "Functions"
.IX Header "Functions"
.SS "debug"
.IX Subsection "debug"
in adding to each of your functions these two lines :
.PP
\&\f(CW$MODE_DEBUG\fR
eval \f(CW$HEADER\fR
.IP "\s-1MODE_DEBUG\s0" 4
.IX Item "MODE_DEBUG"
0 : normal mode
.Sp
1 : as set \-x
.Sp
2 : as set \-xv
.Sp
just put \f(CW$MODE_DEBUG\fR to each fonctions and you could debug on purpose from your main program.
.ie n .IP "eval $HEADER" 4
.el .IP "eval \f(CW$HEADER\fR" 4
.IX Item "eval $HEADER"
HEADER=\*(L"typeset maFonction=\e\*(R"\e${aGras}\e${0}\e${dGras} \-\e\*(L";typeset mf=\e\*(R"\e${maFonction}(\e$1)\e"";
.Sp
as you can see give you function and first arg.
.Sp
maybe be something like \f(CW$0\fR shift then $@ would be better???
.SS "erreur"
.IX Subsection "erreur"
means error in french
.PP
define theses variables on first load :
.PP
OK=0;
KO=1;
ECONT=1;
ESTOP=251;
.PP
usage :
.PP
.Vb 2
\& C<ls my_file>
\& C<erreur $? "mesage" $ECONT "COMMAND">
.Ve
.PP
erreur take the first arg,
.IP "\(bu" 4
if it's equal to 0 it return 0 with the message and \s-1OK\s0
.IP "\(bu" 4
else take the third arg \f(CW$ECONT\fR means continue on error \f(CW$ESTOP\fR means stop on error.
before stoping or continue it start \s-1COMMAND. COMMAND\s0 is optionnal
.PP
\fIhere few exemples\fR
.IX Subsection "here few exemples"
.PP
case1 : no error
.IX Subsection "case1 : no error"
.PP
.Vb 2
\& C<ls my_file>
\& C<erreur $? "presence fichier my_file" $ECONT "ls .">
.Ve
.PP
ls returns 0, so erreur return a message : \fBmyhostname \- 0211 17:21 \- myprogram(myfunction):  presence fichier my_file  \s-1OK\s0\fR
.PP
case2 : continue on error
.IX Subsection "case2 : continue on error"
.PP
same exemple :
.PP
.Vb 2
\& C<ls my_file>
\& C<erreur $? "presence fichier my_file" $ECONT "ls .">
.Ve
.PP
ls return 2 : file not found. so erreur keep the return code of ls I mean 2. then execute the forth arg, in our exemple it's \*(L"ls .\*(R".finally return 2 with the message \fBmyhostname \- 0211 17:28 \- : myprogram(myfunction) presence fichier my_file  \s-1KO\s0\fR
.PP
case3 : no action on error
.IX Subsection "case3 : no action on error"
.PP
the same as explained before without execution of the forth argument because you omited it.
.PP
case4 : stop on error with action
.IX Subsection "case4 : stop on error with action"
.PP
.Vb 2
\& C<ls my_file>
\& C<erreur $? "presence fichier my_file" $ESTOP "ls .">
.Ve
.PP
as you can see I put \f(CW$ESTOP\fR instead of \f(CW$ECONT\fR witch mean please stop on error. The forth arg is executed before the program exit.
.SH "Author"
.IX Header "Author"
Grillon


libsh/libError.sh
#!/usr/bin/env bash

<<=cut

=pod

=encoding utf-8 

=head1 Name

libError

=head1 Description

Error handling and debug library. To use it just source it on the begining of your script.

=head1 Dependencies

no dependencies.

=head1 Functions

=head2 debug

in adding to each of your functions these two lines : 

$MODE_DEBUG
eval $HEADER

=over 4

=item MODE_DEBUG

0 : normal mode

1 : as set -x

2 : as set -xv

just put $MODE_DEBUG to each fonctions and you could debug on purpose from your main program.

=item eval $HEADER

HEADER="typeset maFonction=\"\${aGras}\${0}\${dGras} -\";typeset mf=\"\${maFonction}(\$1)\"";

as you can see give you function and first arg.

maybe be something like $0 shift then $@ would be better???

=back

=head2 erreur

means error in french

define theses variables on first load :

OK=0;
KO=1;
ECONT=1;
ESTOP=251;

usage : 

 C<ls my_file>
 C<erreur $? "mesage" $ECONT "COMMAND">

erreur take the first arg, 

=over 4

=item * 

if it's equal to 0 it return 0 with the message and OK

=item * 

else take the third arg $ECONT means continue on error $ESTOP means stop on error.
before stoping or continue it start COMMAND. COMMAND is optionnal

=back

=head3 here few exemples

=head4 case1 : no error

 C<ls my_file>
 C<erreur $? "presence fichier my_file" $ECONT "ls .">

ls returns 0, so erreur return a message : B<myhostname - 0211 17:21 - myprogram(myfunction):  presence fichier my_file  OK>

=head4 case2 : continue on error

same exemple :

 C<ls my_file>
 C<erreur $? "presence fichier my_file" $ECONT "ls .">

ls return 2 : file not found. so erreur keep the return code of ls I mean 2. then execute the forth arg, in our exemple it's "ls .".finally return 2 with the message B<myhostname - 0211 17:28 - : myprogram(myfunction) presence fichier my_file  KO>

=head4 case3 : no action on error

the same as explained before without execution of the forth argument because you omited it.

=head4 case4 : stop on error with action

 C<ls my_file>
 C<erreur $? "presence fichier my_file" $ESTOP "ls .">

as you can see I put $ESTOP instead of $ECONT witch mean please stop on error. The forth arg is executed before the program exit.

=head1 Author

Grillon


=cut
#variable
OK=0;
KO=1;
ECONT=1;
ESTOP=251;

#commandes
cAwk=${cAwk:=awk}

#titre fonction plus arg1
HEADER="typeset maFonction=\"\${aGras}\${0}\${dGras} -\";typeset mf=\"\${maFonction}(\$1)\"";

#HEADERS
#= verification du nombre d'arguments
#if [ $# -ne 1 ];then erreur $KO "$0 : nombre d'argument incorrects" $ECONT;fi
#= initialisation du debug level
#$MODE_DEBUG

#= titre de la fonction
#typeset maFonction="${aGras}${0}${dGras} -";
#= sous ensemble de la fonction
#typeset mf="${maFonction}($1)"
#- cela peut être remplacé par eval $HEADER


function debug
{
eval $HEADER
if [ $# -ne 1 ];then erreur $KO "$0 : nombre d'argument incorrects" $ECONT;fi
case $1 in
0) MODE_DEBUG=""
	;;
1)	MODE_DEBUG="set -x"
	;;
2)	MODE_DEBUG="set -xv"
	;;
*)	erreur $KO "$0 : argument $1 invalides" $ECONT
	MODE_DEBUG="set -xv"
	return 1
	;;
esac
return 0
}
function erreur
{
$MODE_DEBUG
eval $HEADER
#erreur devient complexe
#arg 1 : numero d'erreur;
#arg 2 : message;
#arg 3 : code erreur(de 251 ` 254), $ECONT pour continuer, $ESTOP pour arret;
#args 4 : fonction ` appeller en cas d'erreur avant sortie

typeset codeRetourE=$1
typeset message=$2
typeset codeRetourS=$3
typeset date=$(date "+%m%d %H:%M")

if [ $codeRetourE -ne 0 ];then
	echo "${aGras} $(uname -n) ${dGras} - $date - : ${tRouge} $message ${fRouge}${tBlanc} KO ${raz}" >&2
	if [ -n "$4" ];then
		typeset action=$4
		eval $action
		ERREUR=$?
	fi
	if [ $codeRetourS -ge 251 ];then
		exit $codeRetourS
	fi
	return $codeRetourE;
else
echo "${aGras} $(uname -n) ${dGras}  - $date - : ${tVert} $message ${fVert}${tNoir} OK ${raz}"
return $OK;
fi
}
libError=1;
